core/amber/src/main/python/core/util/customized_queue/linked_blocking_multi_queue.py

```diff
@@ -0,0 +1,436 @@
+class LinkedBlockingMultiQueue(IKeyedQueue):
+    @inner
+    class Node(Generic[T]):
+        def __init__(self, item: T):
+            self.item = item
+            self.next: Optional[LinkedBlockingMultiQueue.Node[T]] = None
+
+    @inner
+    class SubQueue(Generic[T]):
+        def __init__(self, key: str):
+            self.key: str = key
+            self.priority_group: Optional[LinkedBlockingMultiQueue.PriorityGroup] = None
+            self.put_lock: RLock = RLock()
+            self.count: AtomicInteger = AtomicInteger()
+            self.enabled: bool = True
+            self.head: LinkedBlockingMultiQueue.Node = LinkedBlockingMultiQueue.Node(
+                None
+            )
+            self.last: Optional[LinkedBlockingMultiQueue.Node[T]] = self.head
+
+        def clear(self) -> None:
+            self.fully_lock()
+            try:
+                h: LinkedBlockingMultiQueue.Node[T] = self.head
+                p: LinkedBlockingMultiQueue.Node = h.next
+                while p is not None:
+                    h.next = h
+                    p.item = None
+                    h = p
+                    p = h.next
+                self.head = self.last
+                old_count = self.count.get_and_set(0)
+                if self.enabled:
+                    self.owner.total_count.get_and_dec(old_count)
+            finally:
+                self.fully_unlock()
+
+        def disable(self) -> None:
+            self.fully_lock()
+            try:
+                if not self.enabled:
+                    return
+                self.owner.total_count.dec(self.count.value)
+                self.enabled = False
+            finally:
+                self.fully_unlock()
+
+        def enable(self) -> None:
+            self.fully_lock()
+            try:
+                if self.enabled:
+                    return
+                self.enabled = True
+
+                # potentially unlock waiting polls
+                c = self.count.value
+                if c > 0:
+                    self.owner.total_count.inc(c)
+                    self.owner.not_empty.notify()
+
+            finally:
+                self.fully_unlock()
+
+        def is_enabled(self) -> bool:
+            self.owner.take_lock.acquire()
+            try:
+                return self.enabled
+            finally:
+                self.owner.take_lock.release()
+
+        def enqueue(self, node: LinkedBlockingMultiQueue.Node[T]) -> None:
+            self.last.next = node
+            self.last = node
+
+        def __str__(self) -> str:
+            res = ""
+            h = self.head
+            while h.next is not None:
+                res += h.next.item
+                res += " -> "
+                h = h.next
+            return res
+
+        def size(self) -> int:
+            return self.count.value
+
+        def is_empty(self) -> bool:
+            return self.size() == 0
+
+        def put(self, obj: T) -> None:
+            old_size = -1
```

Comment:
I saw the java implementation is enforcing the obj to be not null, I'm wondering if we should also do this. One reason to support `null` is that python doesn't have an `Optional` wrapper class to use instead of `None`. 

But a minor concern is that the special head pointer is represented using `Node(None)`. so this might or might not be an issue.

Reply:
You have a point. If we want to support None, then when trying to remove it, it could cause issue to the special empty head.

For now since we do not have a use case for None, I decide to drop the support of None. 

Changed to raise ValueError when putting a None value.